mqtt/mqtt-broker/src/configuration.rs

```diff
@@ -0,0 +1,211 @@
+use regex::Regex;
+use serde::{Deserialize, Deserializer};
+use std::path::Path;
+use std::time::Duration;
+
+pub const DEFAULTS: &str = include_str!("config.json");
+
+#[derive(Debug, Deserialize)]
+pub struct InflightMessages {
+    max_count: u32,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct RetainedMessages {
+    max_count: Option<u32>,
+    #[serde(with = "humantime_serde")]
+    expiration: Option<Duration>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct SessionMessages {
+    #[serde(deserialize_with = "humansize")]
+    max_message_size: Option<u64>,
+    max_count: Option<u32>,
+    #[serde(deserialize_with = "humansize")]
+    max_total_space: Option<u64>,
+    when_full: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct SessionPersistence {
+    file_path: String,
+    #[serde(with = "humantime_serde")]
+    time_interval: Option<Duration>,
+    unsaved_message_count: Option<u32>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Session {
+    #[serde(with = "humantime_serde")]
+    expiration: Option<Duration>,
+    messages: SessionMessages,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct BrokerSettings {
+    inflight_messages: InflightMessages,
+    retained_messages: RetainedMessages,
+    session: Session,
+    persistence: Option<SessionPersistence>,
+}
+
+pub fn humansize<'de, D>(deserializer: D) -> Result<Option<u64>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    lazy_static! {
+        static ref SIZE_PATTERN: Regex = Regex::new(r"\s*(\d+)\s*([a-zA-Z]+)\s*")
+            .expect("failed to create new Regex from pattern");
+    }
+
+    let s = String::deserialize(deserializer)?;
+
+    let captures = match SIZE_PATTERN.captures(&s.as_str()) {
+        Some(s) => s,
+        None => return Err(error::<D>(&s.as_str(), &"256kb")),
+    };
+
+    let base = match captures[1].parse::<u64>() {
+        Ok(n) => n,
+        Err(_) => return Err(error::<D>(&captures[1], &"256")),
+    };
+
+    let multiplier = captures[2].to_lowercase();
+    let multiplier = get_multiplier::<D>(multiplier.as_ref())?;
+
+    Ok(Some(base * multiplier))
+}
+
+fn get_multiplier<'de, D>(str: &str) -> Result<u64, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let result = match str {
+        "b" => 1,
+        "kb" => 1024,
+        "mb" => 1024 * 1024,
+        "gb" => 1024 * 1024 * 1024,
+        _ => return Err(error::<D>(str, &"'b', 'kb', 'mb' or 'gb'")),
+    };
+
+    Ok(result)
+}
+
+fn error<'de, D>(unexpected: &str, expected: &str) -> D::Error
+where
+    D: Deserializer<'de>,
+{
+    serde::de::Error::invalid_value(serde::de::Unexpected::Str(&unexpected), &expected)
+}
+
+impl BrokerSettings {
+    pub fn new(path: Option<&Path>) -> Result<Self, ConfigError> {
+        let mut s = Config::new();
+        s.merge(File::from_str(DEFAULTS, FileFormat::Json))?;
+
+        if let Some(path) = path {
+            s.merge(File::from(path))?;
+        }
+
+        s.try_into()
+    }
```

Comment:
I would probably make these two methods instead of using an option.

```rust
impl BrokerSettings {
    pub fn new() -> Result<Self, ConfigError> {
        // return defaults
    }

    pub fn from_file<P: AsRef<Path>>(path: P) -> Result<Self, ConfigError> {
        // merge file with defaults
    }
}
```
Passing an Option as a parameter is kind of an antipattern for situations like this.

Reply:
The caller would do something like this:

```rust
let path: Option<OsString> = "myconfig";
let config = path.map_or(BrokerSetting::new(), BrokerSettings::from_file)?;
```